src/dp/distributions.rs

```diff
@@ -0,0 +1,607 @@
+        assert_eq!(samples, vec!(5, -7, 10, -6, 0, -2, -3, 5, -2, -3));
+        assert_eq!(samples1, vec!(0, 4, 0, 1, 4, -5, 7, 10, 3, -9));
+    }
+
+    #[test]
+    /// Make sure that the distribution created by `create_distribution`
+    /// of `ZCdpDicreteGaussian` is the same one as manually creating one
+    /// by using the constructor of `DiscreteGaussian` directly.
+    fn test_zcdp_discrete_gaussian() {
+        // sample from a manually created distribution
+        let sampler1 =
+            DiscreteGaussian::new(Ratio::<BigUint>::from_integer(BigUint::from(4u8))).unwrap();
+        let mut rng = SeedStreamSha3::from_seed(Seed::from_bytes([0u8; 16]));
+        let samples1: Vec<i8> = (0..10)
+            .map(|_| i8::try_from(sampler1.sample(&mut rng)).unwrap())
+            .collect();
+
+        // sample from the distribution created by the `zcdp` strategy
+        let zcdp = ZCdpDiscreteGaussian {
+            budget: ZCdpBudget::new(Rational::try_from(0.25).unwrap()),
+        };
+        let sampler2 = zcdp
+            .create_distribution(Ratio::<BigUint>::from_integer(1u8.into()))
+            .unwrap();
+        let mut rng2 = SeedStreamSha3::from_seed(Seed::from_bytes([0u8; 16]));
+        let samples2: Vec<i8> = (0..10)
+            .map(|_| i8::try_from(sampler2.sample(&mut rng2)).unwrap())
+            .collect();
+
+        assert_eq!(samples2, samples1);
+    }
+
+    pub fn test_mean<FS: FnMut() -> BigInt>(
+        mut sampler: FS,
+        hyp_mean: f64,
+        hyp_var: f64,
+        alpha: f64,
+        n: u32,
+    ) -> bool {
+        // we test if the mean from our sampler is within the given error margin assuimng its
+        // normally distributed with mean hyp_mean and variance sqrt(hyp_var/n)
+        // this assumption is from the central limit theorem
+
+        // inverse cdf (quantile function) is F s.t. P[X<=F(p)]=p for X ~ N(0,1)
+        // (i.e. X from the standard normal distribution)
+        let probit = |p| Normal::new(0.0, 1.0).unwrap().inverse_cdf(p);
+
+        // x such that the probability of a N(0,1) variable attaining
+        // a value outside of (-x, x) is alpha
+        let z_stat = probit(alpha / 2.).abs();
+
+        // confidence interval for the mean
+        let abs_p_tol = Ratio::<BigInt>::from_float(z_stat * (hyp_var / n as f64).sqrt()).unwrap();
+
+        // take n samples from the distribution, compute empirical mean
+        let emp_mean = Ratio::<BigInt>::new((0..n).map(|_| sampler()).sum::<BigInt>(), n.into());
+
+        (emp_mean - Ratio::<BigInt>::from_float(hyp_mean).unwrap()).abs() < abs_p_tol
+    }
+
+    fn histogram(
+        d: &Vec<BigInt>,
+        bin_bounds: &[Option<(BigInt, BigInt)>],
+        smallest: BigInt,
+        largest: BigInt,
+    ) -> HashMap<Option<(BigInt, BigInt)>, u64> {
+        // a binned histogram of the samples in `d`
+        // used for chi_square test
+
+        fn insert<T>(hist: &mut HashMap<T, u64>, key: &T, val: u64)
```

Comment:
FYI: this helper function could be replaced with `*map.entry(key).or_default() += increment`

Reply:
This is conciser, yes.